plugins/pulp_rpm/plugins/migrations/0043_populate_recommends.py

```diff
@@ -0,0 +1,74 @@
+import gzip
+import logging
+import sys
+
+from pulp.server.db.connection import get_collection
+from pulp.server.db.migrations.lib import utils
+
+if sys.version_info < (2, 7):
+    from xml.etree import ElementTree as ET
+else:
+    from xml.etree import cElementTree as ET
+
+
+_logger = logging.getLogger(__name__)
+
+
+_NAMESPACES = {
+    'common': "http://linux.duke.edu/metadata/common",
+    'rpm': 'http://linux.duke.edu/metadata/rpm',
+}
+
+_HEADER = ('<?mxl version="1.0" encoding="UTF-8"?>\n'
+           '<metadata xmlns="http://linux.duke.edu/metadata/common" '
+           'xmlns:rpm="http://linux.duke.edu/metadata/rpm">\n {}'
+           '</metadata>\n')
+
+
+def migrate_rpm(collection, unit):
+    """
+    Uncompress single rmp unit primary metadata and populate the unit recommends attribute
+    accordingly.
+
+    :param collection: a collection of RPM units
+    :type collection: pymongo.collection.Collection
+    :param unit: the RPM unit being migrated
+    :type unit: dict
+    """
+    primary_xml = unit.get('repodata', {}).get('primary')
+    if not primary_xml:
+        _logger.warning('Encountered an RPM unit without the primary repodata: %s', unit['_id'])
+        return
+
+    _logger.debug('Processing RPM unit %s', unit['_id'])
+    primary_xml = _HEADER.format(gzip.zlib.decompress(primary_xml))
+    root_element = ET.fromstring(primary_xml)
+    delta = {}
+    # the evr+flags fields are actually the attrib attribute of the entry node
+    # <rpm:entry name="foo" epoch="0" version="3.14" release="pi" flags="EQ" />
+    delta['recommends'] = [
+        rpm_entry.attrib for rpm_entry in root_element.iterfind(
+            './common:package/common:format/rpm:recommends/rpm:entry', _NAMESPACES)
+    ]
+    _logger.debug('The unit recommends %', delta['recommends'])
+    if delta['recommends']:
+        collection.update_one({'_id': unit['_id']}, {'$set': delta})
+
+
+def migrate(*args, **kwargs):
+    """
+    Populate the RPM unit recommends attribute.
+    Migration can be safely re-run multiple times.
+
+    :param args: unused
+    :type args: list
+    :param kwargs: unused
+    type kwargs: dict
+    """
+    rpm_collection = get_collection('units_rpm')
+    total_rpm_units = rpm_collection.count()
```

Comment:
besides counting you need to assert that the 'recommends' field is not present otherwise you will migrate even those records which already contain the 'recommends' field in case of import of the unit into upgraded Pulp 2.17

Reply:
could you please elaborate on the reason why recalculating the `recommends` attribute might be bad?

Reply:
I'm worried about consistency; if there were already some `recommends` fields present on the units, we'd better override them rather than relying on what ever was present was sound.

Reply:
OK never mind me, let's do it for the performance sake; the code wouldn't always update the record anyways, as it is now.